src/contract/contract.ts

```diff
@@ -43,91 +58,86 @@ export class TinymanContract {
-      // All of the template variables are ints
-      let value_encoded = encodeVarInt(value);
-      let diff = v.length - value_encoded.length;
+    this.poolLogicSigContractTemplate = poolLogicSig.logic.bytecode;
+    this.templateVariables = poolLogicSig.logic.variables;
+  }
 
-      offset += diff;
+  generateLogicSigAccountForPool(params: {
+    network: SupportedNetwork;
+    asset1ID: number;
+    asset2ID: number;
+  }): LogicSigAccount {
+    const {network, asset1ID, asset2ID} = params;
+    const validatorAppID = getValidatorAppID(network, CONTRACT_VERSION.V1_1);
+    const generateLogicSigAccountForPoolParams: GenerateLogicSigAccountForV1_1PoolParams =
+      {
+        validatorAppID,
+        asset1ID,
+        asset2ID,
+        poolLogicSigContractTemplate: this.poolLogicSigContractTemplate,
+        templateVariables: this.templateVariables
+      };
+
+    return generateLogicSigAccountForV1_1Pool(generateLogicSigAccountForPoolParams);
+  }
+}
 
-      programArray = programArray
-        .slice(0, start)
-        .concat(value_encoded)
-        .concat(programArray.slice(end));
-    }
+export class TinymanContractV2 extends BaseTinymanContract<
+  V1_1ValidatorApp,
+  V2PoolLogicSig
+> {
+  private poolLogicSigContractTemplate: string;
 
-    const program = new Uint8Array(programArray);
+  constructor(validatorApp: V1_1ValidatorApp, poolLogicSig: V2PoolLogicSig) {
+    super(validatorApp, poolLogicSig);
 
-    const lsig = new LogicSigAccount(program);
+    this.poolLogicSigContractTemplate = poolLogicSig.logic.bytecode;
+  }
 
-    return {
-      addr: lsig.address(),
-      program
+  generateLogicSigAccountForPool(params: {
+    network: SupportedNetwork;
+    asset1ID: number;
+    asset2ID: number;
+  }): LogicSigAccount {
+    const {network, asset1ID, asset2ID} = params;
+    const validatorAppID = getValidatorAppID(network, CONTRACT_VERSION.V2);
+    const generateLogicSigAccountForPoolParams: GenerateLogicSigAccountForV2PoolParams = {
+      validatorAppID,
+      asset1ID,
+      asset2ID,
+      poolLogicSigContractTemplate: this.poolLogicSigContractTemplate
     };
+
+    return generateLogicSigAccountForV2Pool(generateLogicSigAccountForPoolParams);
   }
 }
 
-export const tinymanContract = new TinymanContract(
-  ascJson.contracts.validator_app,
-  ascJson.contracts.pool_logicsig
+export const tinymanContract_v1_1 = new TinymanContractV1_1(
+  ascJson_v1_1.contracts.validator_app,
+  ascJson_v1_1.contracts.pool_logicsig
 );
 
-export const validatorAppSchema = tinymanContract.schema;
-
-function encodeVarInt(number) {
-  let buf: number[] = [];
-
-  // eslint-disable-next-line no-constant-condition
-  while (true) {
-    let towrite = number & 0x7f;
-
-    number >>= 7;
-
-    if (number) {
-      buf.push(towrite | 0x80);
-    } else {
-      buf.push(towrite);
-      break;
-    }
-  }
-  return buf;
-}
+export const tinymanContract_v2 = new TinymanContractV2(
+  ascJson_v1_1.contracts.validator_app,
```

Comment:
```suggestion
export const tinymanContract_v2 = new TinymanContractV2(
  // TODO: Use v2 validator app when it's ready
  ascJson_v1_1.contracts.validator_app,
```
I think we better add a `TODO` for this